### ISSUE TITLE
Flex Attention and variable sequence length

### ISSUE THREAD

**Issue by TParcollet** [pytorch-labs/attention-gym#119]
Hello there,

I've been trying to figure out if Flex Attention was variable sequence length friendly but I couldn't make it work (with fixed size, it seems to work well :) ). The idea is that, for speech processing for instance, we commonly have the sequence length dimension varying from one batch to another (and we do padding within the batch). Right now it just crashes with an error. A running toy example would be: 

```python 

import torch
from torch.nn.attention.flex_attention import flex_attention
import torch.nn.functional as F

torch.set_default_device("cuda:0")

# Compile the flex_attention function
flex_attention = torch.compile(flex_attention, dynamic=False)

B = 16
H = 8
S = 200
D = 128

time_base = 0
time_flex = 0

def noop(score, b, h, q_idx, kv_idx):
    return score

import time
for i in range(500):
    S = i*4
    query = torch.randn(
            B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
        )
    key = torch.randn(
            B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
        )
    value = torch.randn(
            B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
        )
        
    if i > 10:
        start_b = time.time()
        with torch.nn.attention.sdpa_kernel(torch.nn.attention.SDPBackend.MATH):
            base = F.scaled_dot_product_attention(query, key, value)
        torch.cuda.synchronize()
        time_base += (time.time() - start_b)

        start_flex = time.time()
        flex = flex_attention(query, key, value, score_mod=noop)
        torch.cuda.synchronize()
        time_flex += (time.time() - start_flex)
    else:
        with torch.nn.attention.sdpa_kernel(torch.nn.attention.SDPBackend.MATH):
            base = F.scaled_dot_product_attention(query, key, value)
        flex = flex_attention(query, key, value, score_mod=noop)

print(time_base)
print(time_flex)
```

**Comment by TParcollet** [comment]
Right, I'm too tired I guess. Thanks for spotting this!

**Comment by drisspg** [comment]
Variable sequence lengths do work, setting dynamic=False will not enable dynamic codegen though.

**Comment by TParcollet** [comment]
@drisspg changing the code to start at one with dynamic = False makes it infinitely long with flexattention (is it recompiling for each new sequence length and caching it?)

Changing dynamic to True just errors out :/
